Sim api

BGluth opened the pull request:
This is the initial version of the runner.

I wrote a decent amount up in a notebook (`Using the Runner.ipynb`), but the general idea is to have something to make it easy to tie data generation with different types of simulations.

Currently it's hooked up to all the data generators (blob, random, and from file) as well as Maharsh's stuff. The goal is for this to be able very easy to expand for when we add new types of parameters/data generation/simulations.

BGluth committed:
Basic logic of simulator

BGluth committed:
Initial wiring of needed functions done for sim

BGluth committed:
Some small bug fixes with the simulator

BGluth committed:
Initial version of simulator notebook

BGluth committed:
Impl intermediate func to call gen_blob

BGluth committed:
Some cleanup

- Black pass
- Removed unused import

BGluth committed:
Impled a data gen type that reads data from a file

BGluth committed:
Fixed rand data gen calling blob data gen

BGluth committed:
Reading in from a datafile now transforms for sim format

BGluth committed:
Added missing set function for file path param

BGluth committed:
Fixed spelling error in sim

BGluth committed:
Moved Maharsh's sim runner logic into sim_run_funcs

- Also made a slight modification so it uses the sims parameter system.

BGluth committed:
Updated required params for run funcs in simulator

BGluth committed:
Updated the sim notebook example to use Maharsh's fed learn logic

BGluth committed:
Python black pass

BGluth committed:
Made the error format for missing params a bit nicer

BGluth committed:
Added a few more parameters specific to fed learning

BGluth committed:
Improved the notebook on using the runner

BGluth committed:
Renamed "Simulator" --> "Runner"

- "Runner" is a totally better name for what it does.
- Also updated usages within files.

crankycoder commented on runner.py:
```diff
@@ -0,0 +1,134 @@
+
+        print("Finished!")
+
+    def set_num_samples(self, num_samples):
```
Let's get rid of most of these setters.  It doesn't make sense that someone would call one or two of the setters without calling all of the setters required to make the Runner get into a valid overall state.

I think it makes sense to just pass in a JSON blob of configuration to the Runner in the constructor, and then call the state validation code in `_sim_has_required_params_for_given_run_func`

BGluth replied:
Agreed. This is probably nicer to work with.

Should I worry about providing an API to create the JSON?

crankycoder replied:
No, I don't think an API to create the JSON is necessary.

crankycoder commented on runner.py:
```diff
@@ -0,0 +1,134 @@
+        self._params[p_key] = val
+        return self
+
+    def _sim_has_required_params_for_given_run_func(
```
Let's call this from the Runner constructor so that we can't ever have a Runner in an invalid startup state.

BGluth replied:
Sure.

crankycoder replied:
`_sim_has_required_params_for_given_run_func` should probably throw an exception if invalid parameters have been passed in.

crankycoder commented on runner.py:
```diff
@@ -0,0 +1,134 @@
+import runner_run_funcs
+
+SIM_TYPE_FED_LEARNING = "fed_learning"
```
These constants don't have a global semantic meaning.  Let's scope these to be class attributes of `Runner`

BGluth replied:
Yeah this makes sense once we get rid of the other classes in the module.

BGluth replied:
I'm wondering now if it actually does make more sense for these to be loose in `runner.py` (or maybe even in a file of their own).

Currently there are multiple files that are dependent on these keys, `runner.py` itself and `runner_run_funcs.py`, and also now any code that's going to be constructing the json. It feels kind of weird to me for these to be defined as statics on `Runner` while in `runner_run_funcs` we don't care at all about the `Runner` class besides these static keys that live on it.

This is kind of nit picky, so IDK. I'd be interested to know your thoughts on this.

BGluth replied:
It's also kind of weird since both `runner.py` and `runner_run_funcs.py` are deps of eachother.

BGluth replied:
There is a dependency issue happening when importing `runner.py`.

`runner` imports `runner_run_funcs`, which needs the static key variables from `Runner` (class), but `runner` hasn't been fully imported yet so it doesn't see it as existing.

I can't see a nice way for these to be static without this issue. There might be (if there is, just post a comment here and I'll make the changes it), but for now I'm going to put the keys loose in their own file.

crankycoder replied:
Can we just cheat and just those two modules together for now?

BGluth replied:
Haha sure. :)

crankycoder commented on runner.py:
```diff
@@ -0,0 +1,134 @@
+P_KEY_NUM_EPOCHS = "num_epochs"
+
+
+class RunFuncAndReqParams:
```
I think we can safely drop this class - it doesn't 'carry it's own weight'.   A code smell for a class that can be removed is that you've got internal state, but no methods attached to the class.

I would replace this with either a simple 2-tuple, or possibly a dict.

BGluth replied:
Yeah I agree. A tuple might make more sense for two fields when it's self contained within a module.

crankycoder commented on runner.py:
```diff
@@ -0,0 +1,134 @@
+    SIM_TYPE_FED_AVG_WITH_DP: RunFuncAndReqParams(
+        runner_run_funcs.run_fed_avg_with_dp, {P_KEY_NUM_LABELS, P_KEY_NUM_FEATURES}
+    ),
+    DATA_GEN_TYPE_DATA_FROM_FILE: RunFuncAndReqParams(
```
Can we split run_func_ltable into two separate lists: 

  * (model, required params) 
  * (data generator, labels)

BGluth replied:
So... I just need some clarification here.

From our meeting, I know we want the way the simulator gets data to feel more "natural" in terms of how this algo. would run in the real world. But I'm still not totally sure what this would look like. At this point should we still even support data generation right in the runner, or only allow reading in from a file?

I think in the example above  `(data generator, labels)` `data generator` is the run function to generate/read-in the data. Is this right? And would `labels` just be the number of labels in the data?

crankycoder replied:
re `(data generator, labels)` - I think so - it's hard to get an idea for the ergonomics of what the API is going to be like without iterating a bunch.  Let's just split the monolithic list for now so that we have smaller parts that are easier to test and instrument so that we can iterate 'cheaply' in terms of effort while we figure out how we want to make this work.

BGluth replied:
Ok I'll give this a try

crankycoder reviewed (commented):
I think if we simplify the structure a little we can get this into shape so we have something that can easily be imported into a thin server module.   I don't want to dwell too much on nits - let's optimize for readability at this point.

BGluth committed:
Removed RunFuncAndReqParams

- Replaced with tuple.

BGluth committed:
Param key constants are now statics in Runner

- NOTE: Currently does not run because of a circular dependencyissue
between runner.py and runner_run_funcs.py. Will fix!

BGluth committed:
Now inits params with json

- Also removed setter funcs.

BGluth committed:
Moved keys to seperate file and gen funcs to runner_run_funcs

- There was a dependency issue between `runner` importing `runner_run_funcs` which imported `runner`.
- Also made sense for the ltable to be in runner_run_funcs.

BGluth committed:
Fixed two small bugs

BGluth committed:
Updated the example notebook to reflect the changes

BGluth commented:
Ok I think I'm ready for another look. @crankycoder 

The only thing that I didn't fully yet implement is moving the data generation ltable stuff to a format like you suggested `(data generator, labels)`. But I did split the sim/datagen run functions into two separate tables.

Also let me know what you think about that dependency issue between `runner.py` and `runner_run_functions.py`.

Also I accidentally rebased onto master when I needed to do a squash, so I ended up including a bunch of commits in the rebase that were not needed.

crankycoder commented on runner.py:
```diff
@@ -0,0 +1,51 @@
+
+    def __init__(self, param_json, sim_type, data_gen_type):
+        self._params = json.loads(param_json)
+        self._run(sim_type, data_gen_type)
```
You should probably rename `_run` to `_init` or just inline the code into the `__init__` method.

crankycoder commented on runner.py:
```diff
@@ -0,0 +1,51 @@
+        ):
+            return
+
+        print('Generating "{}" data...'.format(data_gen_type))
```
The `_run` method seems to have two concerns.  The parts above are initialization related, and the code below is calling `gen_run_func` and `sim_run_func` are actually 'running' the simulation.  Let's split this up so that we have everything initialized at Runner construction time and then executed when run() is invoked.

crankycoder reviewed (commented):
I've got a couple minor nits to split up the `_run` method into an `_init()` (or inline it) and `run()` method, but this looks pretty good to me.

BGluth committed:
Merged `Runner._run` into `Runner.__init__`

- Was kind of small to be it's own function.

BGluth committed:
Checking for missing runner params now raises except

- Previously just did a print and returned a bool.
- Also updated notebook.

BGluth committed:
Condensed some var names to fit validate calls onto one line

- I hope this is ok. I generally don't like condensing names this much,
but when Python Black expands these calls from 1 --> 3 lines, I find it's
not very nice to read.

BGluth committed:
Merged runner_run_funcs & runner_keys (and made keys static again)

- I'll be honest, this is very gross and I really hope this is
temporary. The issue is if we want keys to be static on Runner, then we
will run into a dependency issue if `runner` imports `runner_run_funcs`.
This hacky solution just merges all files together, which is really not
pleasant.

BGluth committed:
Fixed calling a non-existing function for random data gen

BGluth commented:
Ok @crankycoder, the ball is in your court again.

I made all the requested changes, but I'll be dead honest, merging all three files together is very, very, ugly IMHO.

crankycoder reviewed (approved):
This looks good @BGluth! 

Yes - I agree.  The structure isn't optimal yet, but we're trying to do three things here really:

1. implement the federated learning algorithm
2. add a differential privacy version
3. engineer this so that we can deploy to production

I think what you've got now gives us a rough outline of how we can start to think about running simulations against the different implementations.  

What we've got now is a good place to start iterating to get this into a automatically testable state which will get us #3 easier.
